// @wardenprotocol/warden-langchain/dist/warden_tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WardenTool = void 0;
const tools_1 = require("@langchain/core/tools");
class WardenTool extends tools_1.StructuredTool {
    constructor(action, agentkit) {
        super();
        this.action = action;
        this.agentkit = agentkit;
        this.name = action.name;
        this.description = action.description;
        this.schema = action.schema;
    }
    async _call(input) {
        try {
            let args;
            if (this.schema) {
                try {
                    const validatedInput = this.schema.parse(input);
                    args = validatedInput;
                }
                catch (validationError) {
                    args = input;
                }
            }
            else {
                args = input;
            }
            return await this.agentkit.run(this.action, args);
        }
        catch (error) {
            if (error instanceof Error) {
                return `Error executing ${this.name}: ${error.message}`;
            }
            return `Error executing ${this.name}: Unknown error occurred`;
        }
    }
}
exports.WardenTool = WardenTool;
