// @wardenprotocol/warden-agent-kit-core/dist/utils/contracts/constants/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignRequestStatus = exports.KeyRequestStatus = exports.ActionStatus = exports.KeyType = void 0;
var KeyType;
(function (KeyType) {
    // The key type is missing.
    KeyType[KeyType["KEY_TYPE_UNSPECIFIED"] = 0] = "KEY_TYPE_UNSPECIFIED";
    // The key is an ECDSA secp256k1 key.
    KeyType[KeyType["KEY_TYPE_ECDSA_SECP256K1"] = 1] = "KEY_TYPE_ECDSA_SECP256K1";
    // The key is an EdDSA Ed25519 key.
    KeyType[KeyType["KEY_TYPE_EDDSA_ED25519"] = 2] = "KEY_TYPE_EDDSA_ED25519";
})(KeyType || (exports.KeyType = KeyType = {}));
var ActionStatus;
(function (ActionStatus) {
    ActionStatus[ActionStatus["Unspecified"] = 0] = "Unspecified";
    ActionStatus[ActionStatus["Pending"] = 1] = "Pending";
    ActionStatus[ActionStatus["Completed"] = 2] = "Completed";
    ActionStatus[ActionStatus["Revoked"] = 3] = "Revoked";
    ActionStatus[ActionStatus["Timeout"] = 4] = "Timeout";
})(ActionStatus || (exports.ActionStatus = ActionStatus = {}));
var KeyRequestStatus;
(function (KeyRequestStatus) {
    /** KEY_REQUEST_STATUS_UNSPECIFIED - The request is missing the status field. */
    KeyRequestStatus[KeyRequestStatus["KEY_REQUEST_STATUS_UNSPECIFIED"] = 0] = "KEY_REQUEST_STATUS_UNSPECIFIED";
    /**
     * KEY_REQUEST_STATUS_PENDING - The request is waiting to be fulfilled. This is the initial state of a
     * request.
     */
    KeyRequestStatus[KeyRequestStatus["KEY_REQUEST_STATUS_PENDING"] = 1] = "KEY_REQUEST_STATUS_PENDING";
    /** KEY_REQUEST_STATUS_FULFILLED - The request was fulfilled. This is a final state for a request. */
    KeyRequestStatus[KeyRequestStatus["KEY_REQUEST_STATUS_FULFILLED"] = 2] = "KEY_REQUEST_STATUS_FULFILLED";
    /** KEY_REQUEST_STATUS_REJECTED - The request was rejected. This is a final state for a request. */
    KeyRequestStatus[KeyRequestStatus["KEY_REQUEST_STATUS_REJECTED"] = 3] = "KEY_REQUEST_STATUS_REJECTED";
    KeyRequestStatus[KeyRequestStatus["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(KeyRequestStatus || (exports.KeyRequestStatus = KeyRequestStatus = {}));
var SignRequestStatus;
(function (SignRequestStatus) {
    /** SIGN_REQUEST_STATUS_UNSPECIFIED - The request is missing the status field. */
    SignRequestStatus[SignRequestStatus["SIGN_REQUEST_STATUS_UNSPECIFIED"] = 0] = "SIGN_REQUEST_STATUS_UNSPECIFIED";
    /**
     * SIGN_REQUEST_STATUS_PENDING - The request is waiting to be fulfilled. This is the initial state of a
     * request.
     */
    SignRequestStatus[SignRequestStatus["SIGN_REQUEST_STATUS_PENDING"] = 1] = "SIGN_REQUEST_STATUS_PENDING";
    /** SIGN_REQUEST_STATUS_FULFILLED - The request was fulfilled. This is a final state for a request. */
    SignRequestStatus[SignRequestStatus["SIGN_REQUEST_STATUS_FULFILLED"] = 2] = "SIGN_REQUEST_STATUS_FULFILLED";
    /** SIGN_REQUEST_STATUS_REJECTED - The request was rejected. This is a final state for a request. */
    SignRequestStatus[SignRequestStatus["SIGN_REQUEST_STATUS_REJECTED"] = 3] = "SIGN_REQUEST_STATUS_REJECTED";
    SignRequestStatus[SignRequestStatus["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(SignRequestStatus || (exports.SignRequestStatus = SignRequestStatus = {}));
