// @wardenprotocol/warden-agent-kit-core/dist/utils/contracts/constants/known.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KNOWN_SELECTORS = exports.KNOWN_CONTRACTS = void 0;
const chains_1 = require("../../../utils/chains");
const actPrecompileAbi_1 = __importDefault(require("../abi/actPrecompileAbi"));
const erc20Abi_1 = __importDefault(require("../abi/erc20Abi"));
const slinkyPrecompileAbi_1 = __importDefault(require("../abi/slinkyPrecompileAbi"));
const wardenPrecompileAbi_1 = __importDefault(require("../abi/wardenPrecompileAbi"));
const utils_1 = require("viem/utils");
const PRECOMPILE_WARDEN_ADDRESS = "0x0000000000000000000000000000000000000900";
const PRECOMPILE_ACT_ADDRESS = "0x0000000000000000000000000000000000000901";
const PRECOMPILE_SLINKY_ADDRESS = "0x0000000000000000000000000000000000000902";
const PRECOMPILE_CONTRACTS = {
    ACT: {
        abi: actPrecompileAbi_1.default,
        address: PRECOMPILE_ACT_ADDRESS,
    },
    WARDEN: {
        abi: wardenPrecompileAbi_1.default,
        address: PRECOMPILE_WARDEN_ADDRESS,
    },
    SLINKY: {
        abi: slinkyPrecompileAbi_1.default,
        address: PRECOMPILE_SLINKY_ADDRESS,
    },
};
exports.KNOWN_CONTRACTS = {
    [chains_1.wardenChiado.id]: PRECOMPILE_CONTRACTS,
    [chains_1.wardenDevnet.id]: PRECOMPILE_CONTRACTS,
};
exports.KNOWN_SELECTORS = [
    erc20Abi_1.default,
    wardenPrecompileAbi_1.default,
    actPrecompileAbi_1.default,
    slinkyPrecompileAbi_1.default,
].reduce((acc, abi) => {
    const selectors = abi.reduce((acc, item) => {
        if (item.type !== "function") {
            return acc;
        }
        const selector = (0, utils_1.toFunctionSelector)((0, utils_1.formatAbiItem)(item));
        return { ...acc, [selector]: item.name };
    }, {});
    return { ...acc, ...selectors };
}, {});
