// @wardenprotocol/warden-agent-kit-core/dist/utils/chains.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.primaryChain = exports.enabledChains = exports.getChainById = exports.wardenDevnet = exports.wardenChiado = exports.CHAIN_IDS = void 0;
const viem_1 = require("viem");
const chains = __importStar(require("viem/chains"));
// Chain IDs as constants for type safety and reusability
exports.CHAIN_IDS = {
    WARDEN_CHIADO: 10010,
    WARDEN_DEVNET: 12345,
};
/**
 * Warden Protocol's Chiado testnet configuration
 */
exports.wardenChiado = (0, viem_1.defineChain)({
    id: exports.CHAIN_IDS.WARDEN_CHIADO,
    name: "Warden Chiado",
    network: "chiado",
    nativeCurrency: { name: "Ward", symbol: "WARD", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://evm.chiado.wardenprotocol.org"],
        },
    },
});
/**
 * Warden Protocol's development network configuration
 */
exports.wardenDevnet = (0, viem_1.defineChain)({
    id: exports.CHAIN_IDS.WARDEN_DEVNET,
    name: "Warden Devnet",
    network: "devnet",
    nativeCurrency: { name: "Ward", symbol: "WARD", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://evm.devnet.wardenprotocol.org"],
        },
    },
});
// Create a type-safe lookup object for chains by ID
const chainsByID = new Map([
    ...Object.entries(chains).map(([_, chain]) => [chain.id, chain]),
    [exports.wardenChiado.id, exports.wardenChiado],
    [exports.wardenDevnet.id, exports.wardenDevnet],
]);
/**
 * Get chain configuration by chain ID
 */
const getChainById = (id) => chainsByID.get(id);
exports.getChainById = getChainById;
/**
 * List of enabled chains for the application
 */
exports.enabledChains = [
    exports.wardenChiado,
    exports.wardenDevnet,
    chains.mainnet,
    chains.sepolia,
];
/**
 * Primary chain for the application, determined by environment
 */
exports.primaryChain = process.env.PRIMARY_CHAIN === "chiado" ? exports.wardenChiado : exports.wardenDevnet;
